Handle multiple workflows; clean code, flag names

philpennock opened the pull request:
We handled that the latest build might not be the build we want within a
workflow, but did not handle the existence of multiple workflows.

Fixing that, the new filtering logic is much easier to comprehend (IMO).

Changed the command-line flags for sanity, since at least one was
now misnamed.  Kept the old names as aliases for compatibility: no
invocation which previously worked should break with this change.

    -workflow <workflow>       NEW
    -jobname <jobname>         aka -workflow-artifact-build
    -retrieve-count <n>        aka -workflow-depth
    -ignore-later-workflows    NEW

By default, we want the build found for a workflow to be part of the
same workflow invocation as the latest build seen for that workflow,
so that we don't skip back to an older generation.  If instead you
just want "the latest build of that name, in any workflow matching this
name", then use `-ignore-later-workflows`.

Tracing what was happening and where data was used, I've cut down the
number of global variables, and changed URL expansion from `%s`
expansions where I couldn't figure out WTH was happening, to a `${var}`
form of expansion, where the template URLs can now be read and
comprehended.

Cut down on function parameter explosion by introducing a `FilterSet`
type concept, to be passed around, together with the expansion object
for populating URLs.

Added guard against `branch` being explicitly set to empty.

One small golint fix.

philpennock committed:
Handle multiple workflows; clean code, flag names

We handled that the latest build might not be the build we want within a
workflow, but did not handle the existence of multiple workflows.

Fixing that, the new filtering logic is much easier to comprehend (IMO).

Changed the command-line flags for sanity, since at least one was
now misnamed.  Kept the old names as aliases for compatibility: no
invocation which previously worked should break with this change.

    -workflow <workflow>       NEW
    -jobname <jobname>         aka -workflow-artifact-build
    -retrieve-count <n>        aka -workflow-depth
    -ignore-later-workflows    NEW

By default, we want the build found for a workflow to be part of the
same workflow invocation as the latest build seen for that workflow,
so that we don't skip back to an older generation.  If instead you
just want "the latest build of that name, in any workflow matching this
name", then use `-ignore-later-workflows`.

Tracing what was happening and where data was used, I've cut down the
number of global variables, and changed URL expansion from `%s`
expansions where I couldn't figure out WTH was happening, to a `${var}`
form of expansion, where the template URLs can now be read and
comprehended.

Cut down on function parameter explosion by introducing a `FilterSet`
type concept, to be passed around, together with the expansion object
for populating URLs.

Added guard against `branch` being explicitly set to empty.

One small golint fix.

philpennock commented:
(change is only in the commit description, fixed a missing leading hyphen on the last added option)

ydnar commented on cart.go:
```diff
@@ -87,28 +158,46 @@ func main() {
+	// for URL expansion with sane named parameters, and put in everything
+	// we might want too, including filters, in case there are better
+	// URLs we can switch to in future.
+	expansions := Expander{
```
Nifty.

ydnar commented on cart.go:
```diff
@@ -13,60 +13,131 @@ import (
 	flag.StringVar(&circleToken, "token", "", "CircleCI auth token")
 	flag.StringVar(&outputPath, "o", "", "output file `path`")
 	flag.BoolVar(&verbose, "v", false, "verbose output")
 	flag.BoolVar(&dryRun, "n", false, "skip artifact download")
```
While this line/arg didn’t change in this PR, this PR does change the command line interface for `cart`, so making a request:

- [ ] Change `-n` to `-dry-run` (or add an alias)

ydnar commented on cart.go:
```diff
@@ -13,60 +13,131 @@ import (
+	// and "build" as the job within that workflow.
+
+	flag.StringVar(&filter.workflow, "workflow", "", "only consider builds which are part of this workflow")
+	flag.StringVar(&filter.jobname, "jobname", "", "look within workflow for artifacts from this job's name")
```
- [ ] `-jobname` → `-job`

ydnar commented on cart.go:
```diff
@@ -13,60 +13,131 @@ import (
+
+	flag.StringVar(&filter.workflow, "workflow", "", "only consider builds which are part of this workflow")
+	flag.StringVar(&filter.jobname, "jobname", "", "look within workflow for artifacts from this job's name")
+	flag.IntVar(&retrieveBuildsCount, "retrieve-count", defaultRetrieveCount, "how many builds to retrieve")
```
- [ ] `-retrieve-count` → `-count` and `-c`

philpennock replied:
Both that and my name are wrong, because they both imply retrieving N artifacts.  We only ever retrieve from one build.  I'm going with `-search-depth` and no short option because it's not a routine-usage flag.  If we need to override this, then we should change the code default shortly afterwards.

ydnar commented on cart.go:
```diff
@@ -13,60 +13,131 @@ import (
+	flag.StringVar(&filter.jobname, "jobname", "", "look within workflow for artifacts from this job's name")
+	flag.IntVar(&retrieveBuildsCount, "retrieve-count", defaultRetrieveCount, "how many builds to retrieve")
+	// This description is too long; how to make it shorter?
+	flag.BoolVar(&filter.anyFlowID, "ignore-later-workflows", false, "get the last successful workflow/job build, even if of a previous flow-id than the latest we see for that workflow")
```
I can’t figure out what this arg does by looking at the code or the description. Explain it like I’m five?

philpennock replied:
Does the description in the commit message help any?  I can paste it into the code.  If not, I might not be able to explain it.

ydnar commented on cart.go:
```diff
@@ -13,60 +13,131 @@ import (
+	// continue taking the confusingly named option, but map it to the fixed variable.  Similarly for
+	// how the presence of multiple workflows means "workflow-depth" was now a misnomer, and "retrieve-count"
+	// is more accurate.
+	flag.StringVar(&filter.jobname, "workflow-artifact-build", "", "(alias for -workflow-jobname)")
```
Does any code use this legacy arg?

philpennock replied:
Yes, the Makefile of the main private-repo app which earns $$$ invokes `cart` with this flag now, so if we lose the flag immediately then branch builds will fail and create unnecessary pain.

ydnar commented on cart.go:
```diff
@@ -13,60 +13,131 @@ import (
+	// how the presence of multiple workflows means "workflow-depth" was now a misnomer, and "retrieve-count"
+	// is more accurate.
+	flag.StringVar(&filter.jobname, "workflow-artifact-build", "", "(alias for -workflow-jobname)")
+	flag.IntVar(&retrieveBuildsCount, "workflow-depth", defaultRetrieveCount, "(alias for -retrieve-count)")
```
I’d put `-workflow-depth` and `-workflow-artifact-build` next to the preferred arg above, or delete them if we’re confident no code uses it.

philpennock replied:
I'd expect them to be deleted in a month or so, after branches are all merged.  So I'd like them kept out for easier deletion.  I'll add a comment saying "delete after date X".

ydnar reviewed (changes requested):
I was able to run the following (note the missing `-` in front of `slug`) without an error:

`go run cart.go -v repo nbio/cart -branch master -workflow-artifact-build build slug.tgz`

I figure it should fail on mixed flags/args, particularly unrecognized args?

philpennock committed:
Address most review feedback

Add `-dry-run` alias; rename `-jobname` to `-job`, also added shortened
options `-w` and `-j`; renamed -retrieve-count` to `-search-depth` which
is much less misleading.

Mark deprecated flag aliases as such in help output and mark in code
with a safe-to-delete date, to give some time to switch over.  (If
you're using this project and are not within nb.io then please let us
know and mention if this schedule won't work for you.)

Copy explanation of `-ignore-later-workflows` from previous commit
message into comment, and adjust usage text.

Add guard for too many parameters left in argv after option processing,
which catches missing hyphens too.

philpennock commented:
I think that commit addresses every point raised which isn't "no, it needs to be like this" except for the shortened form of `-retrieve-count` where the shortened form led me to parse the term differently and see it as misleading, both in my naming and the shortened form.  So that became `-search-depth` as the flag name and no single-letter variant because it shouldn't normally be needed.

ydnar commented:
@philpennock LGTM. Only thought would be `ignore-later-workflows` → `-eager`?

ydnar commented on cart.go:
```diff
@@ -13,59 +13,140 @@ import (
+	flag.StringVar(&filter.workflow, "w", "", "(short for -workflow)")
+	flag.StringVar(&filter.jobname, "job", "", "look within workflow for artifacts from this build/step/job")
+	flag.StringVar(&filter.jobname, "j", "", "(short for -job)")
+	flag.IntVar(&retrieveBuildsCount, "search-depth", defaultRetrieveCount, "how far back to search in build history")
```
👍

philpennock commented:
Eager is very ambiguous; I think that if you use the tool and have to look up what `-ignore-later-workflows` does, the terms will click and you'll remember it when you see it again several months later, or easily be reminded, whereas `-eager` to me implies something about the speed with which we retrieve, eg setting QoS tags on the TCP connection.

jeffchan commented:
@philpennock  Just wanted to say thank you for your work on this. I ran into the same exact issue yesterday, and was pleasantly surprised to see it fixed today 😄

philpennock commented:
@jeffchan You're welcome.  :)  Please note that I shoved an expiration date on the old workflow flag names, so you've got a couple of months to switch over, unless someone screams.

jeffchan commented:
Thanks for the heads up! I saw that and changed over our usage to the new (and clearer) flags.
